### ISSUE TITLE
Investigate possible whitelisting for dark mode

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#5542]
I wonder if it's possible to somehow inject meta-tags via JS to disable force dark mode:

https://bugs.chromium.org/p/chromium/issues/detail?id=964594
https://bugs.chromium.org/p/chromium/issues/detail?id=925935

Unlikely to actually work, but still worth a try.

**Comment by The-Compiler** [comment]
I've tried this:

```html
<!doctype html>
<html>
    <head>
        <meta name="supported-color-schemes" content="light">
        <meta name="color-scheme" content="light">
    </head>
<body style="forced-color-adjust: none; color-scheme: light">
    booh
</body>
</html>
```

but no luck so far, even with `--qt-flag enable-blink-features=CSSColorScheme,MetaColorScheme`. Not sure what I'm missing...

**Comment by The-Compiler** [comment]
Some progress, thanks to https://bugs.chromium.org/p/chromium/issues/detail?id=1177973#c6 - with:

```html
<meta name="color-scheme" content="dark light">
<style>
  @media (prefers-color-scheme: dark) {
    :root { color: lime }
  }                                                                                                                                                                                                                                           
</style>
<p>This text should be green with forced darkening enabled and a dark preferred OS color scheme</p>
```

and also

```html
<meta name="color-scheme" content="dark light">
<style>
  :root { background-color: white; color: black; }
</style>
<p>This text should be green with forced darkening enabled and a dark preferred OS color scheme</p>
```

the dark mode is actually disabled as long as `-s colors.webpage.preferred_color_scheme dark` is given.

When I use some JS to add it at runtime:

```JS
meta = document.createElement('meta');
meta.name = "color-scheme";
meta.content = "dark light";
document.head.appendChild(meta);
```

that half-works. The background turns white, but the text is also white. It changes when starting to select the text, then it magically turns dark or something. Smells like a Chromium bug.

In either case, I suppose it warrants some more investigation!

**Comment by The-Compiler** [comment]
Hey, this actually works better than I thought it would! It looks like the issue with the text above only happens with a non-conformant HTML file possibly?

Putting this in `config/greasemonkey/whitelist.user.js` seems to work nicely:

```js
// ==UserScript==
// @match https://www.desmos.com/calculator
// ==/UserScript==

const meta = document.createElement('meta');
meta.name = "color-scheme";
meta.content = "dark light";
document.head.appendChild(meta);
```

so I suppose that could actually be turned into some kind of setting, or perhaps even to make `colors.webpage.darkmode.enabled` a per-domain setting (though a bit hackish - only *disabling* it would probably work...).

**Comment by rien333** [comment]
Cool, that works kinda nicely, indeed! 

Not sure if there's a way to improve this, but if this workaround happens to make it into qutebrowser, note that some pages are only drawn in the correct colors after what seems to be a "repaint request" (e.g. a window resize). 

For instance, my banking website initially  has muted grey colors:

![2021-03-03-1614778492](https://user-images.githubusercontent.com/10417027/109813832-2372d480-7c2e-11eb-9658-58481dc50d58.jpg)

But after resizing the window, the text color turns proper black:

![2021-03-03-1614778722](https://user-images.githubusercontent.com/10417027/109813863-2cfc3c80-7c2e-11eb-8a63-bcdcfbd1ae0e.jpg)

**Comment by The-Compiler** [comment]
That looks like what I suspected is a Chromium bug in https://github.com/qutebrowser/qutebrowser/issues/5542#issuecomment-781993145.

I'd hope there's some way for qutebrowser to force such a repaint manually, but I haven't tried yet. It's all quite hacky unfortunately, but I guess still better than nothing.

**Comment by rien333** [comment]
Oddly, the website above (but not all websites per se) loads correctly a first try if I add the following to the header:

```
// ==UserScript==
...
// @run-at document-end
...
// ==/UserScript==
```

This should be the default value of `run-at`, but magically it works if you explicitly specify it.

**Comment by aaronrancsik** [comment]
Hello, I wonder is it possible to make a blacklist instead? I want to specify the sites where forced dark mode is working.
Thank a lot!

**Comment by The-Compiler** [comment]
@aaronrancsik That's not possible as far as I know (except by enabling it globally and then whitelisting everything)

**Comment by aaronrancsik** [comment]
@The-Compiler  It was my first idea too. I will give a try. Thanks!

**Comment by worldpe** [comment]
Would it be possible to make a command to activate darkmode?

**Comment by rien333** [comment]
No, that requires some work on chromium first, see point 21 of the FAQ: https://qutebrowser.org/doc/faq.html

**Comment by groggone** [comment]
This works really^2 well. If people (most especially @aaronrancsik) want to blacklist sites from dark mode, they can just replace "@match" to "@exclude" on the greasemonkey script and it does exactly what you want it to.

```
// ==UserScript==
// @match https://www.desmos.com/calculator
// ==/UserScript==

const meta = document.createElement('meta');
meta.name = "color-scheme";
meta.content = "dark light";
document.head.appendChild(meta);
```

**Comment by The-Compiler** [comment]
FWIW, Qt 6.7 (released a few days ago) adds a [proper run-time toggle](https://codereview.qt-project.org/c/qt/qtwebengine/+/493447) for dark mode. So this might actually be really easy to implement as soon as PyQt 6.7 is out as well.

**Comment by VehementHam** [comment]
> FWIW, Qt 6.7 (released a few days ago) adds a [proper run-time toggle](https://codereview.qt-project.org/c/qt/qtwebengine/+/493447) for dark mode. So this might actually be really easy to implement as soon as PyQt 6.7 is out as well.

Great, gonna wait for that.

**Comment by The-Compiler** [comment]
The hack described here might still be useful for older Qt versions.
